README.md
[![Build Status](https://travis-ci.org/granicus/govdelivery-tms-ruby.svg?branch=master)](https://travis-ci.org/granicus/govdelivery-tms-ruby)

TMS Client
===========
This is a reference Ruby client to interact with the GovDelivery TMS REST API.

Installation
------------
### Using Bundler

```ruby
gem 'govdelivery-tms'
```

### Standalone

```
$ gem install govdelivery-tms
```


Connecting
----------
Loading an instance of `GovDelivery::TMS::Client` will automatically connect to the API to query the available resources for your account.

```ruby
# default api root endpoint is https://tms.govdelivery.com
client = GovDelivery::TMS::Client.new('auth_token', :api_root => 'https://stage-tms.govdelivery.com')
```

Messages
--------

### Loading messages
Sms and Email messages can be retrieved with the `get` collection method.  Messages are paged in groups of 50.  To retrieve another page, used the `next` method.  This method will not be defined if another page is not available.

```ruby
client.sms_messages.get        # get the first page of sms messages
client.sms_messages.next.get   # get the next page of sms messages
```

#### Optional parameters
When loading messages, the following parameters can be passed with requests to change the sort order and number of results returned:

**page_size**: Must be an integer between 1 and 100

```ruby
client.sms_messages.get({page_size: 2})                            # get the first two sms messages
```

**sort_by**: Field by which to sort results. Default: created_at.

```ruby
client.sms_messages.get({sort_by: 'created_at'})                  # get the first page of sms messages, sorted by created_at
```

**sort_order**: Order by which to sort results. Must be ASC or DESC. Default: DESC.

```ruby
client.sms_messages.get({sort_order: 'ASC'})                      # get the first page of sms messages, sorted by created_at DESC
client.sms_messages.get({sort_by: 'body', sort_order: 'ASC'})     # get the first page of sms messages, sorted by body ASC
```

### Sending an SMS Message

```ruby
message = client.sms_messages.build(:body=>'Test Message!')
message.recipients.build(:phone=>'5551112222')
message.recipients.build(:phone=>'5551112223')
message.recipients.build # invalid - no phone
message.post             # true
message.recipients.collection.detect{|r| r.errors } # {"phone"=>["is not a number"]}
# save succeeded, but we have one bad recipient
message.href             # "/messages/sms/87"
message.get              # <GovDelivery::TMS::SmsMessage href=/messages/sms/87 attributes={...}>
```

### Retrieving Inbound SMS Messages
```ruby
client.inbound_sms_messages.get                             # <GovDelivery::TMS::InboundSmsMessages href=/inbound/sms attributes={...}>
inbound_sms = client.inbound_sms_messages.collection.first  # <GovDelivery::TMS::InboundSmsMessage href=/inbound/sms/10041 attributes={...}>
inbound_sms.to                                              # "+15559999999"
inbound_sms.from                                            # "+15005550006"
inbound_sms.attributes                                      # {:from=>"+15005550006", :to=>"+15559999999", :body=>"test", :command_status=>"success", :keyword_response=>"kwidjebo", :created_at=>"2014-11-05T17:15:01Z"}

```

### Sending an Email Message

```ruby
message = client.email_messages.build(:body=>'<p><a href="http://example.com">Visit here</a></p>',
                                      :subject => 'Hey')
message.recipients.build(:email=>'example1@example.com')
message.recipients.build(:email=>'')
message.post             # true
message.recipients.collection.detect{|r| r.errors } # {"email"=>["can't be blank"]}
# save succeeded, but we have one bad recipient
message.href             # "/messages/email/87"
message.get              # <GovDelivery::TMS::EmailMessage href=/messages/email/88 attributes={...}>
```

#### Sending an Email with Macros

```ruby
message = client.email_messages.build(:subject=>'Hello!',
                                      :body=>'<p>Hi <span style="color:red;">[[name]]</span>!</p>',
                                      :macros=>{"name"=>"there"})
message.recipients.build(:email=>'jim@example.com', :macros=>{"name"=>"Jim"})
message.recipients.build(:email=>'amy@example.com', :macros=>{"name"=>"Amy"})
message.recipients.build(:email=>'bill@example.com')
message.post
```

#### From Addresses

From Addresses are read only resources that define which email addresses you
can send an email from and have replies and bounces sent to. From Addresses
also have an associated default display name. If you wish to send a message
from an address that is not your account's default, you will need to specify a
From Address on your Message.

To add or edit From Addresses, you will need to contact your CSC.

```ruby
# Fetch the from_addresses on your account
client.from_addresses.get

# Get the first from address on your account
from_address = client.from_addresses.collection.first

# Lets see what the emails and display name are
puts from_address.from_email
puts from_address.reply_to_email
puts from_address.bound_email
puts from_address.from_name

# Is this from address the account's default?
puts from_address.is_default

# New messages default to using the default from_address of your account
message = client.email_messages.build(:body=>'<p><a href="http://example.com">Visit here</a></p>',
                                      :subject => 'Hey')

# Specifiy a different from_address using the links hash
message.links[:from_address] = from_address.id

# If you want, you can override the form_name on a message
message.from_name = 'Better Name'
message.post
```


### Creating an Email Template

```ruby
template = client.email_templates.build(uuid: 'a-new-template',
                                        subject: 'A templated subject',
                                        body: 'Hi [[name]], this body is from a template.',
                                        macros: {"name"=>"person"})
template.post
```

### Updating an Email Template

*Note: `uuid` cannot be updated.*

```ruby
template = client.email_templates.build(:href => 'template/email/a-new-template')
template.get
template.body = 'Hi [[name]], this body is from a new template.'
template.put
```

### Sending an Email using a Template

Assuming you created `template` above:

```ruby
message = client.email_messages.build
message.links[:email_template] = template.uuid
message.recipients.build(:email=>'jim@example.com', :macros=>{"name"=>"Jim"})
message.recipients.build(:email=>'amy@example.com', :macros=>{"name"=>"Amy"})
message.recipients.build(:email=>'bill@example.com')
message.post
```

### Creating an SMS Template

```ruby
template = client.sms_templates.build(uuid: 'a_new-template',
                                      body: 'Hi, [[name]] this is a tempalte.')
template.post
```

### Updating an SMS Template

*Note: `uuid` cannot be updated.*

```ruby
template = client.sms_templates.build(href: 'template/sms/a_new-template')
template.get
template.body = 'Hi, [[name]] this is a new tempalte.'
template.put
```

### Sending an SMS Message using a Template

Assuming you've created `template` above:

```ruby
message = client.email_messages.build
message.links[:sms_template] = template.uuid
message.recipients.build(:phone=>'5551112222')
message.recipients.build(:phone=>'5551112223')
message.post
```


Webhooks
-------
### POST to a URL when a recipient is blacklisted (i.e. to remove from your list)


```ruby
webhook = client.webhooks.build(:url=>'http://your.url', :event_type=>'blacklisted')
webhook.post # true
```

POSTs will include in the body the following attributes:

  attribute   |  description
------------- | -------------
message_type  | 'sms' or 'email'
status:       |  message state
recipient_url |  recipient URL
messsage_url  |  message URL
error_message |  (failures only)
completed_at  |  (sent or failed recipients only)


Metrics
-------

### Viewing recipients that clicked on a link in an email

```ruby
email_message.get
email_message.clicked.get
email_message.clicked.collection # => [<#EmailRecipient>,...]
```

### Viewing recipients that opened an email

```ruby
email_message.get
email_message.opened.get
email_message.opened.collection # => [<#EmailRecipient>,...]
```

### Viewing a list of statistics for a recipient

```ruby
email_recipient.clicks.get.collection #=> [<#EmailRecipientClick>,...]

email_recipient.opens.get.collection #=> [<#EmailRecipientOpen>,...]
```

Reports
-------

### Message recipient counts
Recipient counts are aggregated across all messages and grouped by message status.

`start` and `end` are required datetime parameters. They must be truncated to the hour and be in ISO 8601 format (YYYY-MM-DDTHH:MM:SSZ).

End dates are exclusive and all results are based on utc time.

#### Email

```ruby
stats = client.reports_messages_email_statistics.get({start: '2017-06-01T10:00:00Z', end: '2017:06:30T18:00:00Z'})     # get email recipient counts for messages sent between 6/1/17 and 6/30/17
stats.recipients
```

#### SMS
```ruby
stats = client.reports_messages_sms_statistics.get({start: '2017-06-01T10:00:00Z', end: '2017:06:30T18:00:00Z'})      # get sms recipient counts for messages sent between 6/1/17 and 6/30/17
stats.recipients
```

Configuring 2-way SMS
---------------------

### Listing Command Types
Command Types are the available commands that can be used to respond to an incoming SMS message.

```ruby
command_types = client.command_types.get
command_types.collection.each do |at|
  puts at.name          # "forward"
  puts at.string_fields # ["url", ...]
  puts at.array_fields  # ["foo", ...]
end
```

### Managing Keywords
Keywords are chunks of text that are used to match an incoming SMS message.

```ruby
# CRUD
keyword = client.keywords.build(:name => "BUSRIDE", :response_text => "Visit example.com/rides for more info")
keyword.post                # true
keyword.name                # 'busride'
keyword.name = "TRAINRIDE"
keyword.put                 # true
keyword.name                # 'trainride'
keyword.delete              # true

# list
keywords = client.keywords.get
keywords.collection.each do |k|
  puts k.name, k.response_text
end
```

### Managing Commands
Commands have a command type and one or more keywords.  The example below configures the system to respond to an incoming SMS message containing the string "RIDE" (or "ride") by forwarding an http POST to `http://example.com/new_url`.  The POST body variables are documented in GovDelivery's [TMS REST API documentation](https://govdelivery.atlassian.net/wiki/display/PM/TMS+Customer+API+Documentation#TMSCustomerAPIDocumentation-Configuring2-waySMS "GovDelivery TMS REST API").

```ruby
# CRUD
keyword = client.keywords.build(:name => "RIDE")
keyword.post
command = keyword.commands.build(
            :name => "Forward to somewhere else",
            :params => {:url => "http://example.com", :http_method => "get"},
            :command_type => :forward)
command.post
command.params = {:url => "http://example.com/new_url", :http_method => "post"}
command.put
command.delete

# list
commands = keyword.commands.get
commands.collection.each do |c|
  puts c.inspect
end
```

### Viewing Command Actions
Each time a given command is executed, a command action is created.

**Note** The actions relationship does not exist on commands that have 0 command actions. Because of this, an attempt to access the command_actions attribute of a
command that has 0 command actions will result in a NoMethodError.

```ruby
# Using the command from above
begin
  command.get
  command_actions = command.command_actions
  command_actions.get
  command_action = command_actions.collection.first
  command_action.inbound_sms_message		# InboundSmsMessage object that initiated this command execution
  command_action.response_body			# String returned by the forwarded to URL
  command_action.status				# HTTP Status returned by the forwarded to URL
  command_action.content_type			# Content-Type header returned by the forwarded to URL
rescue NoMethodError => e
  # No command actions to view
end
```

Logging
-------

Any instance of a [Logger](http://www.ruby-doc.org/stdlib-1.9.3/libdoc/logger/rdoc/Logger.html "Ruby Logger")-like class can be passed in to the client; incoming and outgoing request information will then be logged to that instance.

The example below configures `GovDelivery::TMS::Client` to log to `STDOUT`:

```ruby
logger = Logger.new(STDOUT)
client = GovDelivery::TMS::Client.new('auth_token', :logger => logger)
```

ActionMailer integration
------------------------

You can use TMS from the mail gem or ActionMailer as a delivery method.

Gemfile
```ruby
gem 'govdelivery-tms', :require=>'govdelivery/tms/mail/delivery_method'
```

config/environment.rb
```ruby
config.action_mailer.delivery_method = :govdelivery_tms
config.action_mailer.govdelivery_tms_settings = {
    :auth_token=>'auth_token',
    :api_root=>'https://tms.govdelivery.com'
    }
```


Generating Documentation
------------------------
This project uses [yard](https://github.com/lsegal/yard) to generate documentation.  To generate API documentation yourself, use the following series of commands from the project root:

```
# install development gems
bundle install
# generate documentation
bundle exec rake yard
```
The generated documentation will be placed in the `doc` folder.


Running Tests
-------------
```
bundle install
bundle exec rake
```

Compatibility
-------------
This project is tested and compatible with Ruby >=2.5.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file, as
suggested by [How to use a CHANGELOG](http://keepachangelog.com/).

As of version 3.0.0 this project adheres to [Semantic Versioning](http://semver.org/).

## [3.0.0]
### Changed
- Upgrading to ActiveSupport 5.2.4.3
- Gem now supports only Ruby versions >= 2.5.0 and <= 2.7.1

## [2.8.3]
### Changed
- Move from random point version releases and tie gem version to the version of [TMS](https://tms.govdelivery.com/.version).

## [0.10.1]
### Added
- Replaced ActiveSupport::Inflector.camelize with custom version that does not use acronyms.

## [0.10.0]
### Added
- `from_name` attribute on FromAddress resource, mirroring changes to the TMS API

### Fixed
- Removed ability to POST or PUT to the from_address endpoint, as this feature
  is not supported by the TMS API
- Pinned mime-types dependency to a version that supports jRuby


db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20170324183018) do

  create_table "accounts", force: :cascade do |t|
    t.string   "name",                                  null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "voice_vendor_id",          limit: nil
    t.integer  "email_vendor_id",          limit: nil
    t.integer  "sms_vendor_id",            limit: nil
    t.string   "dcm_account_codes",        limit: 4000
    t.integer  "ipaws_vendor_id",          limit: nil
    t.string   "default_response_text"
    t.string   "sid",                      limit: 32,   null: false
    t.string   "link_encoder",             limit: 30
    t.string   "link_tracking_parameters"
    t.string   "sms_message_type",         limit: 20
  end

  create_table "authentication_tokens", force: :cascade do |t|
    t.integer  "user_id",    limit: nil, null: false
    t.string   "token",                  null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "authentication_tokens", ["token"], name: "i_authentication_tokens_token", unique: true, tablespace: "tsms_indx01"

  create_table "call_scripts", force: :cascade do |t|
    t.integer  "voice_message_id", limit: nil
    t.string   "say_text",         limit: 1000
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "command_actions", force: :cascade do |t|
    t.integer  "command_id",         limit: nil
    t.integer  "inbound_message_id", limit: nil
    t.integer  "status",                         precision: 38
    t.string   "content_type",       limit: 100
    t.string   "response_body",      limit: 500
    t.datetime "created_at",                                    null: false
    t.string   "error_message"
  end

  create_table "commands", force: :cascade do |t|
    t.string   "name"
    t.string   "params",       limit: 4000
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "command_type",              null: false
    t.integer  "keyword_id",   limit: nil
  end

  add_index "commands", ["keyword_id"], name: "index_commands_on_keyword_id", tablespace: "tsms_indx01"

  create_table "email_messages", force: :cascade do |t|
    t.integer  "user_id",                limit: nil
    t.integer  "account_id",             limit: nil,                 null: false
    t.text     "body"
    t.string   "status",                             default: "new", null: false
    t.string   "from_name"
    t.string   "subject",                limit: 400
    t.datetime "completed_at"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "ack"
    t.datetime "sent_at"
    t.boolean  "open_tracking_enabled",  limit: nil
    t.boolean  "click_tracking_enabled", limit: nil
    t.text     "macros"
    t.string   "from_email"
    t.string   "reply_to"
    t.string   "errors_to"
    t.integer  "email_template_id",      limit: nil
    t.integer  "message_type_id",        limit: nil
  end

  add_index "email_messages", ["email_template_id"], name: "i_ema_mes_ema_tem_id", tablespace: "tsms_indx01"
  add_index "email_messages", ["message_type_id"], name: "i_ema_mes_mes_typ_id", tablespace: "tsms_indx01"
  add_index "email_messages", ["user_id", "created_at", "status", "subject", "id"], name: "em_idx4", tablespace: "tsms_indx01"

  create_table "email_recipient_clicks", force: :cascade do |t|
    t.integer  "email_message_id",   limit: nil,  null: false
    t.integer  "email_recipient_id", limit: nil,  null: false
    t.string   "email",              limit: 256,  null: false
    t.string   "url",                limit: 4000, null: false
    t.datetime "clicked_at",                      null: false
    t.datetime "created_at",                      null: false
  end

  add_index "email_recipient_clicks", ["email_message_id", "email_recipient_id", "id", "clicked_at"], name: "erc_idx1", tablespace: "tsms_indx01"
  add_index "email_recipient_clicks", ["email_message_id", "email_recipient_id"], name: "i922bea928d6001e8d90c6daf89654", tablespace: "tsms_indx01"

  create_table "email_recipient_opens", force: :cascade do |t|
    t.integer  "email_message_id",   limit: nil, null: false
    t.integer  "email_recipient_id", limit: nil, null: false
    t.string   "event_ip",                       null: false
    t.string   "email",              limit: 256, null: false
    t.datetime "opened_at",                      null: false
    t.datetime "created_at",                     null: false
  end

  add_index "email_recipient_opens", ["email_message_id", "email_recipient_id", "id", "opened_at"], name: "ero_idx1", tablespace: "tsms_indx01"
  add_index "email_recipient_opens", ["email_message_id", "email_recipient_id"], name: "i46e4b3758023b96cabd6e28d7f0bc", tablespace: "tsms_indx01"

  create_table "email_recipients", force: :cascade do |t|
    t.integer  "message_id",    limit: nil,                 null: false
    t.integer  "vendor_id",     limit: nil
    t.string   "ack"
    t.string   "email"
    t.string   "status",                    default: "new", null: false
    t.string   "error_message", limit: 512
    t.datetime "sent_at"
    t.datetime "completed_at"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.text     "macros"
  end

  add_index "email_recipients", ["message_id", "id"], name: "i_ema_rec_mes_id_id", tablespace: "tsms_indx01"

  create_table "email_templates", force: :cascade do |t|
    t.text     "body",                                 null: false
    t.string   "subject",                              null: false
    t.string   "link_tracking_parameters"
    t.text     "macros"
    t.integer  "user_id",                  limit: nil
    t.integer  "account_id",               limit: nil
    t.integer  "from_address_id",          limit: nil
    t.boolean  "open_tracking_enabled",    limit: nil, null: false
    t.boolean  "click_tracking_enabled",   limit: nil, null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "uuid"
    t.integer  "message_type_id",          limit: nil
  end

  add_index "email_templates", ["account_id", "uuid"], name: "i_ema_tem_acc_id_uui", unique: true, tablespace: "tsms_indx01"

  create_table "email_vendors", force: :cascade do |t|
    t.string   "name",                null: false
    t.string   "worker",              null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "deliveries_sequence"
    t.string   "clicks_sequence"
    t.string   "opens_sequence"
  end

  create_table "from_addresses", force: :cascade do |t|
    t.integer  "account_id",     limit: nil
    t.string   "from_email"
    t.string   "reply_to_email"
    t.string   "bounce_email"
    t.datetime "created_at"
    t.boolean  "is_default",     limit: nil, default: false
    t.string   "from_name"
  end

  create_table "from_numbers", force: :cascade do |t|
    t.integer  "account_id",   limit: nil
    t.string   "phone_number"
    t.datetime "created_at"
    t.boolean  "is_default",   limit: nil, default: false
  end

  create_table "inbound_messages", force: :cascade do |t|
    t.integer  "vendor_id",        limit: nil
    t.string   "caller_phone"
    t.string   "body",             limit: 300
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "vendor_phone",     limit: 100
    t.integer  "keyword_id",       limit: nil
    t.string   "keyword_response"
    t.string   "command_status",   limit: 15
    t.integer  "account_id",       limit: nil
    t.string   "vendor_sid"
  end

  add_index "inbound_messages", ["account_id"], name: "i_inbound_messages_account_id", tablespace: "tsms_indx01"
  add_index "inbound_messages", ["vendor_sid"], name: "i_inbound_messages_vendor_sid", unique: true, tablespace: "tsms_indx01"

  create_table "incoming_voice_messages", force: :cascade do |t|
    t.integer  "from_number_id", limit: nil
    t.string   "play_url",       limit: 512
    t.string   "say_text",       limit: 1000
    t.boolean  "is_default",     limit: nil,                 default: false, null: false
    t.integer  "expires_in",                  precision: 38
    t.datetime "created_at"
  end

  create_table "ipaws_vendors", force: :cascade do |t|
    t.integer  "cog_id",                     limit: nil, null: false
    t.string   "user_id",                                null: false
    t.text     "public_password_encrypted",              null: false
    t.text     "private_password_encrypted",             null: false
    t.binary   "jks",                                    null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "keywords", force: :cascade do |t|
    t.integer  "account_id",    limit: nil, null: false
    t.string   "name",          limit: 160, null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "response_text", limit: 160
  end

  create_table "message_types", force: :cascade do |t|
    t.integer "account_id", limit: nil, null: false
    t.string  "label",                  null: false
    t.string  "code",                   null: false
  end

  create_table "one_time_session_tokens", force: :cascade do |t|
    t.string   "value",                  null: false
    t.integer  "user_id",    limit: nil
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "one_time_session_tokens", ["value"], name: "i_one_tim_ses_tok_val", unique: true, tablespace: "gd_enc_xact_indx01"

  create_table "sessions", force: :cascade do |t|
    t.string   "data",       limit: 4000
    t.string   "session_id",              null: false
    t.datetime "expires_at"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "sessions", ["session_id"], name: "index_sessions_on_session_id", unique: true, tablespace: "gd_enc_xact_indx01"

  create_table "sms_messages", force: :cascade do |t|
    t.string   "body"
    t.datetime "created_at"
    t.datetime "completed_at"
    t.integer  "user_id",         limit: nil
    t.integer  "account_id",      limit: nil,                 null: false
    t.string   "status",                      default: "new", null: false
    t.datetime "sent_at"
    t.integer  "sms_vendor_id",   limit: nil,                 null: false
    t.integer  "sms_template_id", limit: nil
  end

  add_index "sms_messages", ["sms_template_id"], name: "i_sms_messages_sms_template_id", tablespace: "tsms_indx01"

  create_table "sms_prefixes", force: :cascade do |t|
    t.string   "prefix",                    null: false
    t.integer  "account_id",    limit: nil, null: false
    t.integer  "sms_vendor_id", limit: nil, null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "sms_recipients", force: :cascade do |t|
    t.integer  "message_id",      limit: nil,                 null: false
    t.integer  "vendor_id",       limit: nil,                 null: false
    t.string   "phone",                                       null: false
    t.string   "formatted_phone"
    t.string   "ack"
    t.string   "error_message",   limit: 512
    t.datetime "sent_at"
    t.datetime "completed_at"
    t.string   "status",                      default: "new", null: false
    t.datetime "created_at"
  end

  add_index "sms_recipients", ["message_id", "id"], name: "i_sms_recipients_message_id_id", tablespace: "tsms_indx01"

  create_table "sms_templates", force: :cascade do |t|
    t.text     "body",                   null: false
    t.integer  "user_id",    limit: nil
    t.integer  "account_id", limit: nil
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "uuid"
  end

  add_index "sms_templates", ["account_id", "uuid"], name: "i_sms_tem_acc_id_uui", unique: true, tablespace: "tsms_indx01"

  create_table "sms_vendors", force: :cascade do |t|
    t.string   "name",                                                                                                                                                     null: false
    t.string   "username",                                                                                                                                                 null: false
    t.string   "password",                                                                                                                                                 null: false
    t.string   "from_phone",                                                                                                                                               null: false
    t.string   "worker",                                                                                                                                                   null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "help_text",             default: "This service is provided by GovDelivery. If you are a customer in need of assistance, please contact customer support.", null: false
    t.string   "stop_text",             default: "You will no longer receive SMS messages.",                                                                               null: false
    t.string   "default_response_text"
    t.string   "start_text"
  end

  add_index "sms_vendors", ["from_phone"], name: "i_sms_vendors_from_phone", unique: true, tablespace: "tsms_indx01"

  create_table "stop_requests", force: :cascade do |t|
    t.integer  "vendor_id",  limit: nil, null: false
    t.string   "phone",                  null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "account_id", limit: nil
  end

  add_index "stop_requests", ["vendor_id", "account_id", "phone"], name: "i_sto_req_ven_id_acc_id_pho", unique: true, tablespace: "tsms_indx01"

  create_table "users", force: :cascade do |t|
    t.integer  "account_id",         limit: nil,                 null: false
    t.string   "email",                                          null: false
    t.string   "encrypted_password",                             null: false
    t.boolean  "admin",              limit: nil, default: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "users", ["email"], name: "index_users_on_email", unique: true, tablespace: "tsms_indx01"

  create_table "voice_messages", force: :cascade do |t|
    t.string   "play_url",     limit: 512
    t.datetime "created_at"
    t.datetime "completed_at"
    t.integer  "user_id",      limit: nil
    t.integer  "account_id",   limit: nil,                                 null: false
    t.string   "status",                                   default: "new", null: false
    t.datetime "sent_at"
    t.string   "say_text",     limit: 1000
    t.integer  "max_retries",               precision: 38, default: 0,     null: false
    t.integer  "retry_delay",               precision: 38, default: 300,   null: false
    t.string   "from_number"
  end

  create_table "voice_recipient_attempts", force: :cascade do |t|
    t.integer  "voice_message_id",   limit: nil, null: false
    t.integer  "voice_recipient_id", limit: nil, null: false
    t.datetime "completed_at"
    t.string   "ack"
    t.string   "description",        limit: 50
  end

  create_table "voice_recipients", force: :cascade do |t|
    t.integer  "message_id",      limit: nil,                 null: false
    t.integer  "vendor_id",       limit: nil
    t.string   "phone"
    t.string   "formatted_phone"
    t.string   "ack"
    t.string   "error_message",   limit: 512
    t.datetime "sent_at"
    t.datetime "completed_at"
    t.string   "status",                      default: "new", null: false
    t.datetime "created_at"
  end

  add_index "voice_recipients", ["message_id", "id"], name: "i_voi_rec_mes_id_id", tablespace: "tsms_indx01"

  create_table "voice_vendors", force: :cascade do |t|
    t.string   "name",       null: false
    t.string   "username",   null: false
    t.string   "password",   null: false
    t.string   "worker",     null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "webhooks", force: :cascade do |t|
    t.integer  "account_id", limit: nil
    t.string   "event_type", limit: 30
    t.string   "url"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

end
